domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresConnectionManager.java

```diff
@@ -0,0 +1,96 @@
+import org.folio.rest.tools.utils.Envs;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PostgresConnectionManager {
+  private static final Logger LOG = LogManager.getLogger(PostgresConnectionManager.class);
+  private static final int DEFAULT_CACHE_SIZE = 4;
+  private final int cacheSizeLimit;
+  private final List<PgConnection> connectionCache = new ArrayList<>();
+  private String currentTenant = "";
+  private int currentIndex;
+
+  public PostgresConnectionManager () {
+    cacheSizeLimit = Envs.getEnv(Envs.DB_MAXSHAREDPOOLSIZE) != null ?
+        Integer.parseInt(Envs.getEnv(Envs.DB_MAXSHAREDPOOLSIZE)) : DEFAULT_CACHE_SIZE;
+  }
+
+  public void clearCache() {
+    currentTenant = "";
+    currentIndex = 0;
+    connectionCache.clear();
+
+    LOG.debug("Cleared connection manager");
+  }
+
+  public Future<PgConnection> getConnection(Pool pool, String schemaName, String tenantId) {
+    if (!PostgresClient.sharedPgPool) {
+      LOG.debug("Not in shared pool mode");
+      return pool.getConnection().map(PgConnection.class::cast);
+    }
+
+    return getCachedConnection(pool, schemaName, tenantId);
+  }
+
+  private Future<PgConnection> getCachedConnection(Pool pool, String schemaName, String tenantId) {
+    if (tenantId.equals(currentTenant)) {
+      var connection = tryGetCachedConnection();
+      if (connection != null) {
+        return Future.succeededFuture(connection);
+      }
+
+      LOG.debug("Adding to connection cache for tenant {} {}", tenantId, connectionCache.size());
+      return createAndCacheConnection(pool, schemaName, tenantId);
+    }
+
+    clearCache();
+
+    currentTenant = tenantId;
+
+    LOG.debug("Starting new connection cache for tenant {} {}", tenantId, connectionCache.size());
+    return createAndCacheConnection(pool, schemaName, tenantId);
+  }
+
+  private PgConnection tryGetCachedConnection() {
+    if (connectionCache.size() > cacheSizeLimit) {
+      return getCachedConnection();
+    }
+    return null;
+  }
+
+  private Future<PgConnection> createAndCacheConnection(Pool pool, String schemaName, String tenantId) {
+    return pool.getConnection().compose(sqlConnection -> {
+      String sql = PostgresClient.DEFAULT_SCHEMA.equals(tenantId)
+          ? "SET ROLE NONE; SET SCHEMA ''"
+          : ("SET ROLE '" + schemaName + "'; SET SCHEMA '" + schemaName + "'");
+
+      return sqlConnection.query(sql).execute()
+          .map(x -> {
+            connectionCache.add((PgConnection) sqlConnection);
+            return (PgConnection) sqlConnection;
+          });
+    });
+  }
+
+  private PgConnection getCachedConnection() {
+    if (connectionCache.isEmpty()) {
+      throw new IllegalStateException("No connections available");
+    }
+
+    LOG.debug("Returning cache item for tenant {} {}", currentTenant, currentIndex);
+
+    PgConnection nextConnection = connectionCache.get(currentIndex);
+    currentIndex = (currentIndex + 1) % connectionCache.size();
```

Comment:
Any concurrency concern here?

Reply:
I considered that but am making the assumption that since its vertx (single thread event loop) and static we're not having to worry. The shared pool is static in the same way and shared between all clients if I'm remembering correctly. All callers should have the same view of the current tenant id and the cache. At least that's what I want to be true in order for this design to work at all.